framework/src/systems/NonlinearSystem.C

```diff
@@ -462,15 +495,52 @@ NonlinearSystem::computeScalingJacobian()
     for (auto dof_index : dof_indices)
       if (dof_map.local_index(dof_index))
       {
-        // For now we will use the diagonal for determining scaling
-        auto mat_value = scaling_matrix(dof_index, dof_index);
-        inverse_scaling_factors[offset + i] =
-            std::max(inverse_scaling_factors[offset + i], std::abs(mat_value));
+        if (jac_scaling)
+        {
+          // For now we will use the diagonal for determining scaling
+          auto mat_value = (*scaling_matrix)(dof_index, dof_index);
+          jac_inverse_scaling_factors[offset + i] =
+              std::max(jac_inverse_scaling_factors[offset + i], std::abs(mat_value));
+        }
+        if (resid_scaling)
+        {
+          auto vec_value = (*scaling_residual)(dof_index);
+          resid_inverse_scaling_factors[offset + i] =
+              std::max(resid_inverse_scaling_factors[offset + i], std::abs(vec_value));
+        }
       }
   }
 
-  // Get the maximum value across processes
-  _communicator.max(inverse_scaling_factors);
+  if (resid_scaling)
+    _communicator.max(resid_inverse_scaling_factors);
+  if (jac_scaling)
+    _communicator.max(jac_inverse_scaling_factors);
+
+  if (jac_scaling && resid_scaling)
+    for (MooseIndex(inverse_scaling_factors) i = 0; i < inverse_scaling_factors.size(); ++i)
+    {
+      // Be careful not to take log(0)
+      if (!resid_inverse_scaling_factors[i])
+      {
+        if (!jac_inverse_scaling_factors[i])
+          inverse_scaling_factors[i] = 1;
+        else
+          inverse_scaling_factors[i] = jac_inverse_scaling_factors[i];
+      }
+      else if (!jac_inverse_scaling_factors[i])
+        // We know the resid is not zero
+        inverse_scaling_factors[i] = resid_inverse_scaling_factors[i];
+      else
+        inverse_scaling_factors[i] =
+            std::exp(_resid_vs_jac_scaling_param * std::log(resid_inverse_scaling_factors[i]) +
+                     (1 - _resid_vs_jac_scaling_param) * std::log(jac_inverse_scaling_factors[i]));
```

Comment:
`inverse_scaling_factors[i] = _resid_vs_jac_scaling_param *resid_inverse_scaling_factors[i] + (1-_resid_vs_jac_scaling_param)*jac_inverse_scaling_factors[i]`

won't work?

Reply:
No because if say the Jacobian recommends scaling by 1e10 and residual recommends scaling by 1, we would then scale by 5e9 which is hardly a compromise. (This is assuming a choice of 0.5 for the `_resid_vs_jac_scaling_param`)